lib/net.c

```diff
@@ -899,11 +899,13 @@ init_iface(struct gatekeeper_if *iface)
 	uint8_t i;
 	uint8_t num_succ_ports = 0;
 	uint8_t num_slaves_added = 0;
+	uint16_t gatekeeper_max_pkt_burst =
+		get_gatekeeper_conf()->gatekeeper_max_pkt_burst;
 
 	iface->alive = true;
 
 	if (iface->bonding_mode == BONDING_MODE_8023AD &&
-			GATEKEEPER_MAX_PKT_BURST < 2 * iface->num_ports) {
```

Comment:
Once we use separate configurations for the burst size of different functional blocks, do we have to make sure that each block should have a configuration greater than the twice size of the number of ports? I think @cjdoucette may know this better.

Reply:
I don't remember the why, but this constraint may only apply to GK and GT blocks since they drive the vast majority of the packets. We should add a comment explaining the why once we remember it.

Reply:
The documentation (https://doc.dpdk.org/api/rte__eth__bond_8h.html) about this bonding mode says:

- For LACP mode to work the rx/tx burst functions must be invoked at least once every 100ms, otherwise the out-of-band LACP messages will not be handled with the expected latency and this may cause the link status to be incorrectly marked as down or failure to correctly negotiate with peers.
- For optimal performance during initial handshaking the array of mbufs provided to rx_burst should be at least 2 times the slave count size.

So this part of the `if` statement is concerned with the second requirement. The second requirement seems to indicate that this is just for performance reasons during the initial handshake of LACP and is not really a hard requirement. If you look at the RX function for this bonding mode, you can see that received packets are checked to see whether they are LACP packets, and this case is marked as `unlikely` (https://github.com/cjdoucette/dpdk/blob/gatekeeper/drivers/net/bonding/rte_eth_bond_pmd.c#L459).

In my opinion this means that we do not need to explicitly check to make sure this condition (`GATEKEEPER_MAX_PKT_BURST < 2 * iface->num_ports`) is true for these reasons:

- It's not a hard requirement.
- Non-optimal performance of LACP setup is acceptable for our application.
- It's likely that the burst size for most blocks will be greater than twice the number of ports anyway.
- Our application has multiple lcores calling RX burst fairly frequently, definitely at least every 100 ms.

_Can't/shouldn't we keep this verification only for GT and GK block?_

_What does this limit mean when all blocks read from dedicated queues?_

If we decide to make sure this requirement is important enough to be fulfilled, then I think we need to be careful about which blocks fulfill it. Since RX burst is called for a specific queue, we need to make sure we are using the block that would receive the LACP packets to fulfill this requirement.

The queue (and therefore block) that would receive these packets depends on the device. We've run into this issue before:

https://github.com/AltraMayor/gatekeeper/blob/master/lls/main.c#L585-L596

Based on the fact that non-IP frames seem to either be delivered to queue 0 or the first queue configured under RSS for all the NICs we've seen, I think we should make the GK, GT, and LLS blocks ensure their burst size is greater than twice the number of slave ports. Again, that's if we decide this requirement is worth filling.

Right now, there's a timer in the LLS block to make sure RX burst is called every 99 ms to meet the LACP requirement. However, we already have one case where this is not sufficient: when using the Amazon ENA the GK/GT block would need to be the one making sure to call RX burst every 100 ms.

- Doesn't the number of GK or GT blocks change the ideal bust number?

I think the `2 * iface->num_ports` number is somewhat arbitrary but basically represents the ability to receive 1 application packet and 1 LACP control packet per port for each call to RX burst during initial setup. I don't have a source for that, but if it's true then having multiple blocks calling RX burst would reduce this requirement further.

Reply:
Given that the second requirement is only relevant for the **initial** handshake, we can ignore it because the handshake likely takes less than 1s and having Gatekeeper running not optimally for 1s during the initialization is fine.

About the first requirement, I think we should ignore it until it becomes a problem because it's not clear if anyone will ever need LACP on a cloud provider. The only reason we added LACP to Gatekeeper was to couple with IXes. And Gatekeeper servers at IXes are expected to be bare metal like our XIA1 server. Moreover, if someone does find a reason to enable LACP on a cloud provider and GK/GT doesn't read packets for 100ms, Gatekeeper is dead.

Thank you for this careful analysis, @cjdoucette.

Reply:
Thanks @cjdoucette !